bionic/code_references.py

```diff
@@ -0,0 +1,245 @@
+# Unlike Streamlit, we don't import modules to avoid runtime
+# computation and use newer libraries to make the code cleaner.
+# https://github.com/streamlit/streamlit
+
+"""
+This module contains functions which are used to find references in
+your code objects. `get_referenced_objects` function finds those
+references for you given a code object and an additional `CodeContext`
+input, which can be extracted using the `get_code_context` function.
+
+CodeHasher uses these methods as part of hashing an entity function for
+Bionic cache invalidation. Using these methods, CodeHasher can hash the
+references of entity functions (and even their references), so that any
+change in the references are also detected by Bionic to invalidate
+cache.
+"""
+
+import attr
+import dis
+import inspect
+import warnings
+
+from .utils.misc import oneline
+
+
+@attr.s
+class CodeContext:
+    """
+    Holds variable information from function and code object
+    attributes.
+
+    Attributes
+    ----------
+    globals: dict
+        A reference to the dictionary that holds the function’s global
+        variables — the global namespace of the module in which the
+        function was defined.
+    cells: dict
+        A dictionary that tracks all free variables by name and cell
+        variable names by themselves.
+    varnames: dict
+        A dictionary that is used to track local variables by name.
+    """
+
+    globals = attr.ib(default={})
+    cells = attr.ib(default={})
+    varnames = attr.ib(default={})
+
+    def copy(self):
+        return CodeContext(
+            globals=self.globals.copy(),
+            cells=self.cells.copy(),
+            varnames=self.varnames.copy(),
+        )
+
+
+def get_code_context(func) -> CodeContext:
+    code = func.__code__
+
+    # Mapping from variable name to the value if we can resolve it.
+    # Otherwise map to the name.
+    cells = {}
+
+    for var in code.co_cellvars:
+        # Streamlit uses the names of the variable instead of the value
+        # for cell variables. I suspect this is so because it is hard
+        # to find the value of a cell variable. It doesn't mean we
+        # don't hash these values for Smart Caching. These cell
+        # variables are also present in ``code.co_consts``` which we
+        # hash separately.
+        cells[var] = var
+
+    if code.co_freevars:
+        assert len(code.co_freevars) == len(func.__closure__)
+        for freevar, cell in zip(code.co_freevars, func.__closure__):
+            cells[freevar] = cell.cell_contents
+
+    varnames = {}
+    if inspect.ismethod(func):
+        varnames = {"self": func.__self__}
+
+    return CodeContext(globals=func.__globals__, cells=cells, varnames=varnames)
+
+
+def get_referenced_objects(code, context):
+    """
+    Attempts to return all objects referenced externally by a code object. In
+    some cases, these objects are
+    - replaced with strings: when we can't find a variable or when the reference
```

Comment:
Just realized this: doesn't returning strings as placeholders also introduce the possibility of collisions? Like if we change this code

    x = call_func() # Suppose we can't following this reference

to 

    x = "call_func"

Can a situation like this produce the same hash?

WDYT about wrapping these placeholder strings in a special class? This might also solve the issue of using the `tos` variable for both reference values and reference names.

Reply:
I don't think an example like that will result in possible collisions because we also hash the `func.__code__.co_code` variable which is the bytecode hash of the function itself. Those two functions will have different `co_code`.

We can hash those strings in a separate class, but we will have to wait till we hash classes for it to work. I'll leave a TODO to change it once we have support for hashing classes.